File: scraper/scraper.py
import json
import os
import psycopg
import requests
import time

from datetime import datetime, timedelta
from dotenv import load_dotenv
from typing import Dict, Any, List


load_dotenv()

conn = psycopg.connect(os.getenv("DATABASE_URL"))


PERIOD_API_URL = f"https://{os.getenv('BASE_API_URL')}/locations/5b50c589f3eeb609b36a87eb/periods/?date=%s"
API_URL = f"https://{os.getenv('BASE_API_URL')}/locations/5b50c589f3eeb609b36a87eb/menu?period=%s&date=%s"
MEAL_TYPES = set(["breakfast", "lunch", "dinner"])

START_DATE = datetime(2025, 9, 16).date()
END_DATE = datetime(2025, 9, 17).date()


def add_item(item_data: Dict[str, Any]) -> None:
    nutrients = {
        nutrient_data["name"].split(" (")[0].strip(): nutrient_data["valueNumeric"].strip() + nutrient_data["uom"].strip()
        for nutrient_data in item_data["nutrients"]
    }
    filters = [
        filter_data["name"].strip()
        for filter_data in item_data["filters"]
        if filter_data["type"] == "label"
    ]

    try:
        with conn.cursor() as cur:
            cur.execute(
                "INSERT INTO items VALUES (%s, %s, %s, %s, %s, %s);",
                (item_data["name"],
                 item_data["desc"].strip() if item_data["desc"] is not None else None,
                 item_data["portion"].strip() if item_data["portion"] is not None else None,
                 item_data["ingredients"].strip().replace("^", ""), json.dumps(nutrients), json.dumps(filters))
            )

        conn.commit()
    except psycopg.IntegrityError:
        conn.rollback()
    except Exception as e:
        conn.rollback()
        print('Error adding item:', e)

def add_menu(date: str, meal: str, location: str, status: str) -> int:
    try:
        with conn.cursor() as cur:
            cur.execute(
                "INSERT INTO menus (date, meal, location, status) VALUES (%s, %s, %s, %s) RETURNING ID;",
                (date, meal, location, status)
            )
            value = cur.fetchone()[0]
    
        conn.commit()
        return value
    except Exception as e:
        conn.rollback()
        print('Error adding menu:', e)
        return None

def add_menu_items(menu_id: int, items: List[str]) -> None:
    try:
        with conn.cursor() as cur:
            cur.executemany(
                """INSERT INTO menu_items VALUES (%s, %s)
                   ON CONFLICT (menu_id, item_name) DO NOTHING;""",
                [(menu_id, item) for item in items]
            )

        conn.commit()
    except Exception as e:
        conn.rollback()
        print('Error adding menu items:', e)


def scrape_menus(date: str, refresh_menus: bool = False) -> bool:
    """
    Scrapes breakfast, lunch, and dinner menus for a given date.
    Inserts items, locations, and menus into the database.
    
    Args:
        date: Date string in YYYY-MM-DD format
        refresh_menus: If True, clears existing menu data for the date before scraping
    """

    response = requests.get(
        PERIOD_API_URL % (date),
        headers={
            "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/139.0.0.0 Safari/537.36",
        },
        timeout=5
    )

    data = response.json()
    
    if refresh_menus:
        if "periods" not in data:
            raise Exception(f"No periods found for {date}")

        try:
            with conn.cursor() as cur:
                cur.execute("""
                    DELETE FROM menu_items 
                    WHERE menu_id IN (
                        SELECT id FROM menus WHERE date = %s
                    );
                """, (date,))
                
                cur.execute("DELETE FROM menus WHERE date = %s;", (date,))
            
            conn.commit()
            print(f"Cleared existing menu data for {date}")
        except Exception as e:
            conn.rollback()
            print(f'Error clearing menu data for {date}:', e)

    # Closed
    if not data["periods"]:
        print(f"No periods found for {date}")
        for meal_type in MEAL_TYPES:
            add_menu(date, meal_type, None, "closed")
    
        return

    meals = {
        period["id"]: period["slug"]
        for period in data["periods"]
        if period["slug"] in MEAL_TYPES
    }
    
    for meal_hash, meal_type in meals.items():
        response = requests.get(
            API_URL % (meal_hash, date),
            headers={
                "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/139.0.0.0 Safari/537.36",
            },
            timeout=5
        )

        data = response.json()
        
        for location_data in data["period"]["categories"]:
            items = []
            for item_data in location_data["items"]:
                item_data["name"] = item_data["name"].strip()
                add_item(item_data)
                items.append(item_data["name"])
            
            menu_id = add_menu(date, meal_type, location_data["name"], "open")

            add_menu_items(menu_id, items)     
            
            print(f"Added menu for {date} {meal_type} {location_data['name']}")

def main():
    with conn.cursor() as cur:
        cur.execute("""
            DELETE FROM menus
            WHERE date = (SELECT MAX(date) FROM menus);
        """)

        cur.execute("""
            SELECT DISTINCT date FROM menus;
        """)
        scraped_dates = {row[0] for row in cur.fetchall()}

    conn.commit()

    all_dates = {START_DATE + timedelta(days=i) for i in range((END_DATE - START_DATE).days)}
    missing_dates = sorted(all_dates - scraped_dates)
   
    for current_date in missing_dates:
        print(f'Obtaining menus for {current_date.strftime("%Y-%m-%d")}.')
        scrape_menus(current_date.strftime("%Y-%m-%d"))
        current_date += timedelta(days=1)
        time.sleep(5)


if __name__ == "__main__":
    main()

    conn.close()

File: scraper/scraper.go
package scraper

import (
	"encoding/json"
	"fmt"
	"log"
	"sync"
	"time"

	"github.com/joho/godotenv"
	"github.com/valyala/fasthttp"
)

const (
	APIURL    = "https://api.dineoncampus.com/v1/location/5b50c589f3eeb609b36a87eb/periods/%s?platform=0&date=%s"
	UserAgent = "Mozilla/5.0 (iPhone; CPU iPhone OS 16_6 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/16.6 Mobile/15E148 Safari/604.1"
)

var mealTypes = map[string]string{
	"66bf79f3351d5300dd055257": "breakfast",
	"66bf7d21e45d430859cf99b2": "lunch",
	"66bf7d21e45d430859cf99b8": "dinner",
}

type Response struct {
	Closed bool `json:"closed"`
	Menu   struct {
		Periods struct {
			Categories []struct {
				Name  string `json:"name"`
				Items []struct {
					Name        string `json:"name"`
					Desc        string `json:"desc"`
					Portion     string `json:"portion"`
					Ingredients string `json:"ingredients"`
					Calories    string `json:"calories"`
					Nutrients   []struct {
						Name          string `json:"name"`
						Value_Numeric string `json:"value_numeric"`
						UOM           string `json:"uom"`
					} `json:"nutrients"`
					Filters []struct {
						Name string `json:"name"`
						Type string `json:"type"`
					} `json:"filters"`
				} `json:"items"`
			} `json:"categories"`
		} `json:"periods"`
	} `json:"menu"`
}

func get(date string, meal string, wg *sync.WaitGroup) {
	defer wg.Done()
	url := fmt.Sprintf(APIURL, meal, date)

	req := fasthttp.AcquireRequest()
	defer fasthttp.ReleaseRequest(req)

	req.SetRequestURI(url)
	req.Header.SetMethod("GET")
	req.Header.Set("User-Agent", UserAgent)

	resp := fasthttp.AcquireResponse()
	defer fasthttp.ReleaseResponse(resp)

	err := fasthttp.DoTimeout(req, resp, 30*time.Second)
	if err != nil {
		log.Printf("error fetching URL: %v", err)
		return
	}

	var data Response
	if err := json.Unmarshal(resp.Body(), &data); err != nil {
		log.Printf("error parsing JSON: %v", err)
		return
	}

	if err := Save(&data, date, mealTypes[meal]); err != nil {
		log.Printf("error saving data: %v", err)
		return
	}
}

func Scrape(date string) error {
	var wg sync.WaitGroup

	for meal := range mealTypes {
		wg.Add(1)
		go get(date, meal, &wg)
	}

	wg.Wait()
	return nil
}

func Save(response *Response, date string, meal string) error {
	err := godotenv.Load(".env")
	if err != nil {
		log.Fatal("Error loading .env file")
		return nil
	}
	// *********** IMPLEMENT DATABASE***********

	data := *response
	if data.Closed {
		log.Printf("Closed")
		return nil
	}

	for _, restaurant := range data.Menu.Periods.Categories {
		for _, item := range restaurant.Items {
			for _, nutrient := range item.Nutrients {
				//log.Printf("Nutrient: %s %s %s", nutrient.Name, nutrient.Value_Numeric, nutrient.UOM)
				_ = nutrient
			}
			for _, filter := range item.Filters {
				//log.Printf("Filter: %s %s", filter.Name, filter.Type)
				_ = filter
			}
			log.Print(meal)
		}
	}

	return nil
}
